0016-pre-shared-key/0016-pre-shared-key.md

```diff
@@ -0,0 +1,277 @@
+
+### Public Headers
+
+The public headers and their data are formatted as below:
+
+```http
+PSK/1.0
+Nonce: fpwpAhlN588
+Key: hmac-sha-256
+Encryption: aes-256-gcm 58EowcXBk3qBIvJ0kmvdCh
+Header: data_everyone_can_see
+
+...
+```
+
+The public headers come after a single status line, which must be `PSK/1.0`. Public headers are visible to all parties transmitting the packet, so they SHOULD contain only the bare minimum required for the intended parties to decrypt and validate the private headers.
+
+PSK payments use an HMAC of the PSK details as a fulfillment, to prove that the contents are unmodified. The pre-shared secret key is the key used in the HMAC.
+
+The data attached to these headers is an encrypted binary blob. This encrypted blob contains the private PSK headers.
+
+Despite the fact that connectors can read the public headers' data, connectors should not be required to do so, nor should they rely on PSK details being present on all transfers. This information is only intended for the initial sender and final receiver of a payment.
+
+The decryption key is derived from the pre-shared secret key, and the AES-256-GCM initialization vector is set to the value of the `Nonce` header (`fpwpAhlN588` in the above example). The nonce MUST be generated with cryptographically-secure randomness. **If the nonce is reused with the same shared secret, it could leak unencrypted data or allow money to be stolen by malicious parties.** The AES-256-GCM authentication tag used for decryption is attached to the `Encryption` header (`58EowcXBk3qBIvJ0kmvdCh` in the above example).
+
+| Header | Value |
+|--------|-------|
+| `Nonce` | _(Required)_ A [base64url-encoded](https://en.wikipedia.org/wiki/Base64#URL_applications) nonce, used to generate ensure uniqueness of the PSK data and as an initialization vector (IV) for AES-256-GCM encryption. The nonce **MUST** be 16 bytes, and **MUST** be generated with cryptographically-secure randomness. |
+| `Encryption` | _(Required)_ Supported values are `aes-256-gcm <AUTH_TAG>` and `none`. If it is set to `aes-256-gcm`, then private headers and application data will be AES-256-GCM encrypted, and `<AUTH_TAG>` will be the authentication tag returned by the cipher. If it is set to `none`, then private headers and application data will be in cleartext. This cleartext will still be appended to the public headers after an empty line. If the value is neither `aes-256-gcm` nor `none`, the receiver MUST reject the incoming payment with an `S06` error. |
+| `Key` | _(Optional)_ The algorithm by which the receiver generates the shared secret. If this value is set to anything but `hmac-sha-256`, the receiver MUST reject the incoming payment with an "S??" (error code to be determined) error. |
```

Comment:
Receivers should support the behavior I described so that if someone using a later version sends something that's unsupported they'll get the right error. Right now we are using `hmac-sha-256` and though it's not necessary for the sender to specify, I think it would make sense for the receiver to accept it.